### DISCUSSION TITLE
"The component initialization has failed" error.

### DISCUSSION THREAD

**Discussion by Rufio2031** [saucecontrol/PhotoSauce#169] [Q&A] [answered]
I recognize this likely isn't a problem with the library itself, but we're having trouble identifying the root cause of the error:
"The component initialization has failed".

We believe the issue is somewhere withing the realm of converting a CMYK image to RGB using the default ColorSpaceMode in ProcessImage(). We are using Windows Server 2022 and Windows 11, and believe we have the updated codecs but also aren't super confident in that realm. We're not sure if we need to enable something, or install something, or if there are windows configurations that need to be applied.

Here are a couple call stacks.

at PhotoSauce.MagicScaler.WicTransforms.AddPixelFormatConverter(PipelineContext ctx, Boolean allowPbgra)\r\n   at PhotoSauce.MagicScaler.Transforms.MagicTransforms.AddNormalizingFormatConverter(PipelineContext ctx, Boolean lastChance)\r\n   at PhotoSauce.MagicScaler.MagicImageProcessor.buildPipeline(PipelineContext ctx, Boolean closedPipeline)\r\n   at PhotoSauce.MagicScaler.MagicImageProcessor.ProcessImage(Stream imgStream, Stream outStream, ProcessImageSettings settings)\r\n

at PhotoSauce.MagicScaler.WicTransforms.AddColorspaceConverter(PipelineContext ctx)\r\n   at PhotoSauce.MagicScaler.Transforms.MagicTransforms.AddColorspaceConverter(PipelineContext ctx)\r\n   at PhotoSauce.MagicScaler.MagicImageProcessor.buildPipeline(PipelineContext ctx, Boolean closedPipeline)\r\n   at PhotoSauce.MagicScaler.MagicImageProcessor.ProcessImage(Stream imgStream, Stream outStream, ProcessImageSettings settings)\r\n

If there is any guidance on this we would be grateful to hear it. Attaching an image that we are able to repro with in our environments.
![Channel_digital_image_CMYK_color (1)](https://github.com/user-attachments/assets/efc3dea1-e871-445b-afbd-6415b90cead9)

**Comment by saucecontrol** [comment]
Yeah, that does look like a COM error coming from the WIC component for CMYK->RGB conversion.  You might look for an inner COM exception and see if there's an HRESULT with more info.

Are you saying you see the error on both Windows 11 *and* on Server 2022?

**Reply by Rufio2031** [reply]
Actually, it looks like it's working correctly on Windows 11. I miscomprehended communication.

However, it is failing on several servers ranging from Server 2016, 2019, and 2022. All these environments are Core.
The HRESULT that we are seeing is **0x88982F8B**. Unfortunately, we are not receiving any inner exception. The extent of what we are finding is: "The component initialization has failed. (0x88982F8B)" with the aforementioned amended call stacks.

**Reply by iamcarbon** [reply, marked as answer]
Windows Server Core is a stripped down OS, and does not include the Windows Imaging Component used by MagicScaler. You'll need to run the full version of Windows Server.

**Reply by saucecontrol** [reply]
Yep, that'd do it.  If you're only getting exceptions when attempting to process CMYK images, it sounds like you're already running with the native codec plugins, otherwise I'd expect you to see errors related to missing codecs.

It is possible to handle this scenario on Windows Server Core using the (undocumented) support for [Little-CMS](https://github.com/mm2/Little-CMS) if you're unable to run Server Standard.  That would mean:
1) Building `lcms2.dll` and including it with your app's binaries ([vcpkg](https://github.com/microsoft/vcpkg) makes this relatively easy).
2) Setting the `PhotoSauce.MagicScaler.EnableWindowsLcms` [AppContext switch](https://learn.microsoft.com/en-us/dotnet/api/system.appcontext.setswitch) to tell the pipeline to look for LCMS.

**Reply by Rufio2031** [reply]
We were able to determine that our server was missing `WindowsCodecsExt.dll`, potentially among other missing pieces. We briefly tried, but failed to manually install that dll separately just to see if we'd have any luck. Now knowing the situation we're in, we're comfortable putting the issue aside for now. Thank you both for your insight!

I will aim to try the Little-CMS approach at a later date. I appreciate the detailed steps and links to follow!
